plugins/backend-trace-forwarder/src/Cardano/BM/Backend/TraceForwarder.lhs

```diff
@@ -161,3 +181,94 @@ data TraceForwarderBackendFailure
 instance Exception TraceForwarderBackendFailure
 
 \end{code}
+
+\subsubsection{Asynchronously reading log items from the queue and sending them to an acceptor.}
+\begin{code}
+spawnDispatcher :: ToJSON a => TraceForwarderMVar a -> IO (Async.Async ())
+spawnDispatcher tfMVar = Async.async $ processQueue
+ where
+  processQueue :: IO ()
+  processQueue = do
+    currentTF <- readMVar tfMVar
+    -- Read the next log item from the queue. If the queue is still empty -
+    -- blocking and waiting for the next log item.
+    nextItem <- atomically $ TBQ.readTBQueue (tfQueue currentTF)
+    -- Try to write it to the handle. If there's a problem with connection,
+    -- this thread will initiate re-establishing of the connection and
+    -- will wait until it's established.
+    sendItem tfMVar nextItem
+    -- Continue...
+    processQueue
+
+-- | Try to send log item to the handle.
+sendItem :: ToJSON a => TraceForwarderMVar a -> LogObject a -> IO ()
+sendItem tfMVar lo = do
+  tfHandle <$> readMVar tfMVar >>= \case
+    Nothing -> do
+      -- There's no handle, initiate the connection.
+      establishConnection 1 1 tfMVar
+      -- Connection is re-established, try to send log item.
+      sendItem tfMVar lo
+    Just h ->
+      try (BSC.hPutStrLn h $! encodeUtf8 (hostname . loMeta $ lo)) >>= \case
+        Right _ ->
+          -- Hostname was written to the handler successfully,
+          -- try to write serialized LogObject.
+          try (BSC.hPutStrLn h $! bs) >>= \case
+            Right _ ->
+              -- Everything is ok, LogObject was written to the handler.
+              return () -- Everything is ok, LogObject was written to the handler.
+            Left (_e :: IOException) -> do
+              reConnectIfQueueIsAlmostFull
+              threadDelay 10000
+              sendItem tfMVar lo
+        Left (_e :: IOException) -> do
+          reConnectIfQueueIsAlmostFull
+          threadDelay 10000
+          sendItem tfMVar lo
+ where
+  (_, bs) = jsonToBS lo
+
+  jsonToBS :: ToJSON b => b -> (Int, BS.ByteString)
+  jsonToBS a =
+    let bs' = BL.toStrict $ encode a
+    in (BS.length bs', bs')
+
+  -- Handle is bad, it looks like the connection is broken.
+  -- Check if the queue is almost full.
+  reConnectIfQueueIsAlmostFull = do
+    currentTF <- readMVar tfMVar
+    currentQueueSize <- atomically $ TBQ.lengthTBQueue (tfQueue currentTF)
+    when (queueIsAlmostFull currentQueueSize) $ do
+      -- The queue is almost full, it means that log items will be dropped soon.
+      -- Initiate re-establishing of connection.
+      modifyMVar_ tfMVar $ \be -> return $ be { tfHandle = Nothing }
```

Comment:
if the handle is already open, doesn't it need to be closed here?

Reply:
Fixed.